packages/playwright-core/src/server/socksClientCertificatesInterceptor.ts

```diff
@@ -21,10 +21,49 @@ import fs from 'fs';
 import tls from 'tls';
 import stream from 'stream';
 import { createSocket } from '../utils/happy-eyeballs';
-import { globToRegex, isUnderTest } from '../utils';
+import { globToRegex, isUnderTest, ManualPromise } from '../utils';
 import type { SocksSocketClosedPayload, SocksSocketDataPayload, SocksSocketRequestedPayload } from '../common/socksProxy';
 import { SocksProxy } from '../common/socksProxy';
 import type * as channels from '@protocol/channels';
+import { debugLogger } from '../utils/debugLogger';
+
+class ALPNCache {
+  private _cache = new Map<string, ManualPromise<string>>();
+
+  get(host: string, port: number, success: (protocol: string) => void, error: (error: Error) => void): void {
```

Comment:
Why take two callbacks instead of making this an async method? Also, instead of rejecting I'd return `http/1.1`, so that the regular error handling logic from the proxy took place instead of something special.